lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -205,135 +210,223 @@ def _on_list_backups_action(self, event: ActionEvent) -> None:
                     )
-                    if (
-                        self.get_service_status(output)
-                        == BackupServiceState.SNAPSHOT_RESTORE_ERROR_INDEX_NOT_CLOSED
-                    ):
-                        # The .opensearch-sap-log-types-config, for example does not show
-                        # in the _cat/indices but does block a restore operation.
-                        to_close = output["error"]["reason"].split("[")[2].split("]")[0]
-                        o = self._request("POST", f"{index}/_close")
-                        logger.debug(
-                            f"_restore_and_try_close_indices_if_needed: request closing {index} returned {o}"
-                        )
-                        closed_idx.add(index)
-                        # We still have non closed indices, try again
+                    not_closed = {
+                        idx
+                        for idx, state in close_resp.get("indices", {}).items()
+                        if state == IndexStateEnum.OPEN
+                    }
+                    if not close_resp.get("acknowledged", False) or not_closed:
                         raise Exception()
-                    # If status returns success, ensure all indices have been updated
-                    # otherwise, raise an assert error and retry
-                    shard_status = output["snapshot"]["shards"]
-                    assert shard_status["total"] == shard_status["successful"]
-        except RetryError:
-            logger.error("_restore_and_try_close_indices_if_needed: fail all retries")
-        return output, closed_idx
+        except RetryError as e:
+            if not close_resp.get("acknowledged", False):
+                raise OpenSearchRestoreFailedClosingIdxError(
+                    f"_close_indices_if_needed: failed closing {to_close}"
+                )
+            elif not_closed:
+                raise OpenSearchRestoreFailedClosingIdxError(
+                    f"_close_indices_if_needed: failed closing {not_closed}"
+                )
+            raise OpenSearchRestoreError(e)
+        return to_close
+
+    def _restore(
+        self, backup_id: int, backup_indices: List[str], wait_for_completion: bool = False
+    ) -> Dict[str, Any]:
+        """Runs the restore and processes the response according to wait_for_completion flag."""
+        backup_indices = self._list_backups()[backup_id]["indices"]
+        output = self._request(
+            "POST",
+            f"_snapshot/{S3_REPOSITORY}/{backup_id}/_restore?wait_for_completion={str(wait_for_completion).lower()}",
+            payload={
+                "indices": ",".join(
+                    [f"-{idx}" for idx in INDICES_TO_EXCLUDE_AT_RESTORE & set(backup_indices)]
+                ),
+                "ignore_unavailable": False,
+                "include_global_state": False,
+                "include_aliases": True,
+                "partial": False,
```

Comment:
I kept the "partial", because I want it to be more explicit for a future traveller:
```
        output = self._request(
            "POST",
            f"_snapshot/{S3_REPOSITORY}/{backup_id}/_restore?wait_for_completion={str(wait_for_completion).lower()}",
            payload={
                "indices": ",".join(
                    [f"-{idx}" for idx in INDICES_TO_EXCLUDE_AT_RESTORE & set(backup_indices)]
                ),
                "partial": False,  # It is the default value, but we want to avoid partial restores
            },
        )
```